File: benchmarks/Csharp/Benchmarks/Richards/TaskControlBlock.cs
namespace AreWeFastYet;

delegate TaskControlBlock ProcessFunction(Packet work, RBObject word);

sealed class TaskControlBlock : TaskState
{
    private Packet input;
    private readonly ProcessFunction function;
    private readonly RBObject handle;

    internal TaskControlBlock(TaskControlBlock aLink, int anIdentity, int aPriority, Packet anInitialWorkQueue, TaskState anInitialState, ProcessFunction aBlock, RBObject aPrivateData)
    {
        Link = aLink;
        Identity = anIdentity;
        Priority = aPriority;
        input = anInitialWorkQueue;
        IsPacketPending = anInitialState.IsPacketPending;
        IsTaskWaiting = anInitialState.IsTaskWaiting;
        IsTaskHolding = anInitialState.IsTaskHolding;
        function = aBlock;
        handle = aPrivateData;
    }

    public int Identity { get; }
    public TaskControlBlock Link { get; }
    public int Priority { get; }

    public TaskControlBlock AddInputAndCheckPriority(Packet packet, TaskControlBlock oldTask)
    {
        if (NO_WORK == input)
        {
            input = packet;
            IsPacketPending = true;
            if (Priority > oldTask.Priority)
                return this;
        }
        else
        {
            input = Append(packet, input);
        }
        return oldTask;
    }

    public TaskControlBlock RunTask()
    {
        Packet message = NO_WORK;
        if (IsWaitingWithPacket)
        {
            message = input;
            input = message.Link;
            if (NO_WORK == input)
                SetRunning();
            else            
                SetPacketPending();
        }
        return function(message, handle);
    }
}

File: benchmarks/C++HJ/richards/TaskControlBlock.cpp
#include "TaskState.h"
#include <memory>
#include "Packet.h"
#include <functional>
#include <utility>
#include <iostream>
#include <memory>

using namespace std;

namespace richards {
    class TaskControlBlock : public TaskState, public enable_shared_from_this<TaskControlBlock> {
        private: 

            shared_ptr<TaskControlBlock> _link;
            int _identity;
            int _priority;
            shared_ptr<Packet> _input;
            function<shared_ptr<TaskControlBlock>(shared_ptr<Packet> work, shared_ptr<RBObject> word)> _function;
            shared_ptr<RBObject> _handle;

        public:

            TaskControlBlock(shared_ptr<TaskControlBlock> aLink, int anIdentity, int aPriority, shared_ptr<Packet> anInitialWorkQueue,
                    const shared_ptr<TaskState>& anInitialState, function<shared_ptr<TaskControlBlock>(shared_ptr<Packet> work, shared_ptr<RBObject> word)> aBlock,
                    shared_ptr<RBObject> aPrivateData) {
                _link = move(aLink);
                _identity = anIdentity;
                _priority = aPriority;
                _input = move(anInitialWorkQueue);
                setPacketPending(anInitialState->isPacketPending());
                setTaskWaiting(anInitialState->isTaskWaiting());
                setTaskHolding(anInitialState->isTaskHolding());
                _function = move(aBlock);
                _handle = move(aPrivateData);
            }

            int getIdentity() const {
                return _identity; 
            }

            shared_ptr<TaskControlBlock> getLink()  { 
                return _link; 
            }

            int getPriority() const {
                return _priority; 
            }

            shared_ptr<TaskControlBlock> addInputAndCheckPriority(shared_ptr<Packet> packet, shared_ptr<TaskControlBlock> oldTask) {
                if (NO_WORK == _input) {
                    _input = packet;
                    setPacketPending(true);
                    if (_priority > oldTask->getPriority()) { 
                        return shared_from_this();
                    }
                } else {
                    _input = append(packet, _input);
                }
                return oldTask;
            }

            shared_ptr<TaskControlBlock> runTask() {
                shared_ptr<Packet> message;
                if (isWaitingWithPacket()) {
                    message = _input;
                    _input = message->getLink();
                    if (NO_WORK == _input) {
                        running();
                    } else {
                        packetPending();
                    }
                } else {
                    message = NO_WORK;
                }
                return _function(message, _handle);
            }
    };

}